R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+    asList = stack$as_list,
+    uniqueList = function() {
+      unique(stack$as_list())
+    }
+  )
+}
+
+# (Used for `unique_envirStack()` only. Do not use directly!)
+envirStackUni_ <- function() {
+  map <- fastmap()
+  stack <- faststack()
+  list(
+    push = function(env) {
+      key <- env$envKey
+      if (!map$has(key)) {
+        # mark the key as _seen_
+        map$set(key, TRUE)
+        # add the env
+        stack$push(env)
+      }
+    },
+    uniqueList = stack$as_list
+  )
+}
+# Use to retrieve unique environments (eg: `tq$parent()`)
+envirStackUnique <- function() {
+  stack <- envirStack()
+  count <- 0
+  list(
+    push = function(env) {
+      count <<- count + 1
+      if (count == 500) {
+        # convert the current stack to a `envirStackUni_()`
+        newStack <- envirStackUni_()
+        walk(stack$asList(), newStack$push)
+        stack <<- newStack
+      }
+      stack$push(env)
+    },
+    uniqueList = function() {stack$uniqueList()}
+  )
+}
+
+
+
+
+
+
+# Retrieve all attributes that can be manually set
+# ?attr
+# Note that some attributes (namely ‘class’, ‘comment’, ‘dim’,
+# ‘dimnames’, ‘names’, ‘row.names’ and ‘tsp’) are treated specially
+# and have restrictions on the values which can be set.
+safeAttrValues <- function(x) {
+  badElAttrs <- c("class", "comment", "dim", "dimnames", "names", "row.names", "tsp")
+  attrVals <- attributes(x)
+  attrVals[badElAttrs] <- NULL
+  attrVals
+}
+
+# Convert a list to an environment and keep class and attribute information
+safeListToEnv <- function(x, newClass = NULL) {
+  xList <- x
+
+  ret <- list2env(xList, new.env(parent = emptyenv()))
+
+  attrVals <- safeAttrValues(xList)
+  walk2(names(attrVals), attrVals, function(attrName, attrValue) {
+    attr(ret, attrName) <<- attrValue
+  })
+
+  class(ret) <- c(newClass, class(xList), "environment")
+  ret
+}
+
+# Convert an environment to a list and keep class and attribute information
+safeEnvToList <- function(x, oldClass = NULL) {
+  xEnv <- x
+  ret <- as.list.environment(xEnv, all.names = TRUE)
+
+  attrVals <- safeAttrValues(xEnv)
+  walk2(names(attrVals), attrVals, function(attrName, attrValue) {
+    attr(ret, attrName) <<- attrValue
+  })
+
+  class(ret) <- setdiff(class(x), c(oldClass, "environment"))
+  ret
+}
+
+
+# Do not export to encourage direct use of `tagQuery()`
+asTagEnv <- function(x) {
+  if (isTagQuery(x)) {
+    return(asTagEnv(x$root()))
+  }
+
+  if (!isTagEnv(x)) {
+    if (!isTag(x)) {
+      # force all methods to send in tags, lists / tagLists are not allowed
+      stop("`asTagEnv()` can only accept tag envs or tag objects. It does not accept `lists()` or `tagLists()`")
+    }
+  }
+  asTagEnv_(x, parent = x$parent)
+}
+asTagEnv_ <- function(x, parent = NULL, seenMap = envirMap()) {
+  isTagVal <- isTag(x)
+  isTagEnvVal <- isTagEnv(x)
+
+  if (isTagVal || isTagEnvVal) {
+    if (!isTagEnvVal) {
+      xList <- x
+      x <- safeListToEnv(xList, "htmltools.tag.env")
```

Comment:
Is there a particular reason not to hard-code `"htmltools.tag.env"` inside of `safeListToEnv()`?

Reply:
I was trying to keep `safeListToEnv()` a generic function. Like something that `staticimports` could use.

It also makes it more explicit as to what is happening.

Reply:
(leaving as is)